### PR TITLE
`make_quantile_score_candidates` proof

### LINKED ISSUE #700: `make_quantile_score_candidates` proof

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #700

### CODE REVIEW

**Comment by Shoeboxam** [comment]
Current dependencies on/for this PR:
* main
  * **PR #696** <a href="https://app.graphite.dev/github/pr/opendp/opendp/696" target="_blank">Graphite</a> 
    * **PR #645** <a href="https://app.graphite.dev/github/pr/opendp/opendp/645" target="_blank">Graphite</a> 
      * **PR #604** <a href="https://app.graphite.dev/github/pr/opendp/opendp/604" target="_blank">Graphite</a> 
        * **PR #699** <a href="https://app.graphite.dev/github/pr/opendp/opendp/699" target="_blank">Graphite</a> 
        * **PR #687** <a href="https://app.graphite.dev/github/pr/opendp/opendp/687" target="_blank">Graphite</a> 
          * **PR #688** <a href="https://app.graphite.dev/github/pr/opendp/opendp/688" target="_blank">Graphite</a> 
            * **PR #689** <a href="https://app.graphite.dev/github/pr/opendp/opendp/689" target="_blank">Graphite</a> 
              * **PR #702** <a href="https://app.graphite.dev/github/pr/opendp/opendp/702" target="_blank">Graphite</a> 
                * **PR #703** <a href="https://app.graphite.dev/github/pr/opendp/opendp/703" target="_blank">Graphite</a>  👈
              * **PR #690** <a href="https://app.graphite.dev/github/pr/opendp/opendp/690" target="_blank">Graphite</a> 
                * **PR #512** <a href="https://app.graphite.dev/github/pr/opendp/opendp/512" target="_blank">Graphite</a> 
                * **PR #692** <a href="https://app.graphite.dev/github/pr/opendp/opendp/692" target="_blank">Graphite</a> 
                  * **PR #697** <a href="https://app.graphite.dev/github/pr/opendp/opendp/697" target="_blank">Graphite</a> 
                  * **PR #514** <a href="https://app.graphite.dev/github/pr/opendp/opendp/514" target="_blank">Graphite</a> 
      * **PR #691** <a href="https://app.graphite.dev/github/pr/opendp/opendp/691" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/opendp/opendp/703?utm_source=stack-comment).

File: python/src/opendp/transformations.py [hunk 1]
```diff
@@ -1921,6 +1921,10 @@ def make_quantile_score_candidates(
     * Input Metric:   `MI`
     * Output Metric:  `LInfDiffDistance<usize>`
     
+    **Proof Definition:**
+    
+    [(Proof Document)](https://docs.opendp.org/en/latest/proofs/rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.pdf)
```
Comment (PaulineMauryL) [inline python/src/opendp/transformations.py:1926]: The link does not work yet.
Is it expected as it is not merged ?

File: rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+\end{proof}
+
+Take any two elements $X, X'$ in the \texttt{input\_domain} and any pair $(\din, \dout)$, 
+where \din\ has the associated type for \texttt{input\_metric} and \dout\ has the associated type for \texttt{output\_metric}.
+Assume $X, X'$ are \din-close under \texttt{input\_metric} and that $\texttt{stability\_map}(\din) \leq \dout$. 
+
+\begin{align*}
+    \dout &= \max\limits_{X \sim X'} d_{IDD}(s, s') &&\text{where } s = \function(X)\\
+    &= \max\limits_{X \sim X'} \max\limits_{ij} |(s_i - s'_i) - (s_j - s'_j)| &&\text{by definition of }\rustdoc{metrics/struct}{LInfDiffDistance} \\
+    &\leq 2 \max\limits_{X \sim X'} \max\limits_{i} |s_i - s'_i| &&\text{$s$ is not monotonic; take looser bound}\\
```
Comment (PaulineMauryL) [inline rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex:243]: I do not understand how to get to these 2 lines (?)

File: rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+\end{proof}
+
+Take any two elements $X, X'$ in the \texttt{input\_domain} and any pair $(\din, \dout)$, 
+where \din\ has the associated type for \texttt{input\_metric} and \dout\ has the associated type for \texttt{output\_metric}.
+Assume $X, X'$ are \din-close under \texttt{input\_metric} and that $\texttt{stability\_map}(\din) \leq \dout$. 
+
+\begin{align*}
+    \dout &= \max\limits_{X \sim X'} d_{IDD}(s, s') \\
+    &= \max\limits_{X \sim X'} \max\limits_{ij} |(s_i - s'_i) - (s_j - s'_j)| &&\text{by definition of }\rustdoc{metrics/struct}{LInfDiffDistance} \\
+    &\leq 2 \max\limits_{X \sim X'} \max\limits_{i} |s_i - s'_i| &&\text{$s$ is not monotonic; take looser bound}\\
```
Comment (PaulineMauryL) [inline rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex:299]: and here also (maybe I just don't have the background for it). is there a link ?

File: rust/src/transformations/quantile_score_candidates/pseudocode/make_quantile_score_candidates.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    def function(arg: List[TIA]) -> List[usize]:
+        return compute_score(arg, candidates, alpha_numer, alpha_denom, size_limit)
+
+    if input_domain.size is not None:
+        def stability_map(d_in: u32) -> usize:
+            return TOA.inf_cast(d_in // 2).inf_mul(4).inf_mul(alpha_denom)
+    else:
+        abs_dist_const: usize = max(alpha_numer, alpha_denom.inf_sub(alpha_numer))
+        sup_dist_const: usize = abs_dist_const.inf_mul(2)
+        stability_map = new_stability_map_from_constant(sup_dist_const, QO=usize)
```
Comment (PaulineMauryL) [inline rust/src/transformations/quantile_score_candidates/pseudocode/make_quantile_score_candidates.py:34]: 2 · d_in · max(αnum, αden − αnum)

multiply by d_in missing?

**Review by PaulineMauryL — APPROVED** [review]

File: rust/src/transformations/quantile_score_candidates/pseudocode/make_quantile_score_candidates.py (hunk 4 above)
Comment (Shoeboxam) [reply rust/src/transformations/quantile_score_candidates/pseudocode/make_quantile_score_candidates.py:34]: The d_in isn't missing there, it's just hidden! `new_stability_map_from_constant(C)` gives you a stability map of `d_in * C`

File: rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex (hunk 2 above)
Comment (Shoeboxam) [reply rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex:243]: Will update to make this clearer. The gist is that it is always true, by the triangle inequality, that 
```
|(s_i - s'_i) - (s_j - s'_j)| <= |(s_i - s'_i)| + |- (s_j - s'_j)| <= |(s_i - s'_i)| + |- (s_i - s'_i)| <= 2 \cdot |s_i - s'_i|
```

File: rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+
+Take any two elements $X, X'$ in the \texttt{input\_domain} and any pair $(\din, \dout)$, 
+where \din\ has the associated type for \texttt{input\_metric} and \dout\ has the associated type for \texttt{output\_metric}.
+Assume $X, X'$ are \din-close under \texttt{input\_metric} and that $\texttt{stability\_map}(\din) \leq \dout$. 
+
+\begin{align*}
+    \dout &= \max\limits_{X \sim X'} d_{IDD}(s, s') &&\text{where } s = \function(X)\\
+    &= \max\limits_{X \sim X'} \max\limits_{ij} |(s_i - s'_i) - (s_j - s'_j)| &&\text{by definition of }\rustdoc{metrics/struct}{LInfDiffDistance} \\
+    &\leq 2 \max\limits_{X \sim X'} \max\limits_{i} |s_i - s'_i| &&\text{$s$ is not monotonic; take looser bound}\\
+    &\leq 2 \sum_j^{\din} \max\limits_{Z_{j} \sim Z_{j+1}} \max\limits_{i} |s_{i,j} - s_{i,j+1}| &&\text{by path property where } d_{Sym}(Z_i, Z_{i+1}) = 1, X = Z_0 \text{ and } Z_\din = X' \\
```
Comment (PaulineMauryL) [inline rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex:244]: it is a detail but maybe both in same order 
X = Z_0 \text{ and }  X'= Z_\din
or
Z_0 = X \text{ and } Z_\din = X'

File: rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+This formula matches the stability map in the case where the dataset size is unknown.
+
+
+\subsubsection{Known Size Stability}
+\label{sec:known-size}
+Now consider the case where the dataset size is known.
+
+\begin{lemma}
+    \label{single-change-stab}
+    If $d_{CO}(X, X') \le 1$, then $d_\infty(\function(X), \function(X')) \le 2$.
```
Comment (PaulineMauryL) [inline rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex:258]: \le 2 * alpha_{den}

File: rust/src/transformations/quantile_score_candidates/pseudocode/make_quantile_score_candidates.py (hunk 4 above)
Comment (PaulineMauryL) [reply rust/src/transformations/quantile_score_candidates/pseudocode/make_quantile_score_candidates.py:34]: yes, makes sense, thanks

File: python/src/opendp/transformations.py (hunk 1 above)
Comment (PaulineMauryL) [reply python/src/opendp/transformations.py:1926]: ok

File: rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex (hunk 6 above)
Comment (Shoeboxam) [reply rust/src/transformations/quantile_score_candidates/make_quantile_score_candidates.tex:258]: Thanks!

**Review by andrewvyrros — APPROVED** [review]
A couple of formatting nits.

Otherwise, approved subject to `contrib`.

**Review by andrewvyrros — APPROVED** [review]

**Review by andrewvyrros — APPROVED** [review]

**Pull request merged**